DREAM/Optimization/tsgOptimizationUtils.hpp

```diff
@@ -105,6 +118,45 @@ inline ObjectiveFunction makeObjectiveFunction(const int num_dimensions, const O
     };
 }
 
+/*! \ingroup OptimizationUtil
+ * \brief Generic non-batched gradient function signature.
+ *
+ * Accepts a single input \b x and returns the gradient at the point \b x.
+ */
+using GradientFunctionSingle = std::function<std::vector<double>(const std::vector<double> &x_batch)>;
```

Comment:
How does this work in DREAM? Do we return a vector or do we pass in a pre-allocated vector or pointer, which avoids the allocation.

Reply:
In DREAM, we pass a pre-allocated vector to the function. Should we do this here as well? I assumed RVO would help avoid extra allocations.

Reply:
It's better to be consistent and pass in a pre-allocated vector.

Reply:
Done. I changed the signature of `GradientFunctionSingle` and `ProjectionFunctionSingle` to have a pre-allocated vector.